lib/jack_tokenizer.rb

```diff
@@ -1,34 +1,47 @@
 class JackTokenizer
   def initialize(io)
-    @raw_tokens = io.read.split(/\s(?=(?:[^"]|"[^"]*")*$)/)
+    @input = io.read
+    @index = 0
   end
 
   def has_more_tokens?
-    !@raw_tokens.empty?
+    @index < @input.length
   end
 
   def advance
-    @current_token = @raw_tokens.shift
+    if @input[@index] == '"'
+      next_index = @input.index('"', @index + 1) + 1
+    else
+      next_index = @input.index(" ", @index)
+    end
+
+    @current_token = @input[@index...next_index]
+
+    if SYMBOL_TOKENS.include?(@current_token)
+      @token_type = :SYMBOL
+    elsif KEYWORD_TOKENS.include?(@current_token)
+      @token_type = :KEYWORD
+    elsif DIGITS.include?(@current_token[0])
+      @token_type = :INT_CONST
+    elsif @current_token.start_with?('"')
+      @token_type = :STRING_CONST
+    else
+      @token_type = :IDENTIFIER
+    end
+
+    if next_index.nil?
+      @index = @input.length
+    else
+      @index = next_index + 1
+    end
   end
 
   KEYWORD_TOKENS = %w[class method function constructor int boolean char void var static field let do if else while return true false null this]
   SYMBOL_TOKENS =  %w[{ } ( ) [ ] . , ; + - * / & | < > = ~]
+  DIGITS = %w[0 1 2 3 4 5 6 7 8 9]
```

Comment:
This is good, but since all we do with `DIGITS` is call `#include?` on it, we could use any other object which implements that predicate. For example:

```suggestion
  DIGITS = '0'..'9'
```

Reply:
Asking for the sake of discussion, which one is more efficient?

Reply:
Let’s take a look:

```ruby
>> require 'benchmark/ips'
=> true
>> Benchmark.ips do |x|
     digits = %w[0 1 2 3 4 5 6 7 8 9]
     
     x.report('array') do
       digits.include?('7')
     end
     
     digits = '0'..'9'
     
     x.report('range') do
       digits.include?('7')
     end
     
     x.compare!
   end

Warming up --------------------------------------
               array   315.148k i/100ms
               range   315.567k i/100ms
Calculating -------------------------------------
               array      3.106M (± 2.6%) i/s -     15.757M in   5.076042s
               range      3.148M (± 0.9%) i/s -     15.778M in   5.012718s

Comparison:
               range:  3147898.1 i/s
               array:  3106479.2 i/s - same-ish: difference falls within error
```

So: the range is a tiny bit faster, but there’s not much in it. It’s a very small collection in any case (only ten elements) so I would be surprised if we saw much of a performance difference either way; if it was a million elements then perhaps we’d be able to see more of a gap.

If we look at [MRI’s implementation](https://github.com/ruby/ruby/blob/v3_0_2/range.c#L1532-L1577) of `Range#include?`, we can see that it has some special efficient cases (mostly involving delegation to [`#cover?`](https://rubydoc.info/stdlib/core/Range:cover%3F)) for situations where one or both bounds of the range are numbers, strings or `nil`; perhaps unsurprisingly, [TruffleRuby’s implementation](https://github.com/oracle/truffleruby/blob/vm-21.3.0/src/main/ruby/truffleruby/core/range.rb#L358-L369) does something similar. The reason for this should be fairly intuitive: if I ask you whether the range `100000..1000000000` includes the value `123456`, you can figure that out without having to make a list of almost a billion numbers and check through all of them, just by comparing `123456` with the start and end of the range to see whether it falls between them (which is explicitly what `#cover?` does). But the start and end of a range can be any object, not just a number, so the general fallback is to [call `super`](https://github.com/ruby/ruby/blob/v3_0_2/range.c#L1537) (i.e. [`Enumerable#include?`](https://rubydoc.info/stdlib/core/Enumerable:include%3F)) which literally does iterate through every element of the range comparing it to the value, and that’ll be roughly as slow as checking every element of an array of the same size.

So the tldr is: ranges are generally more efficient than arrays because they use less memory (you only have to store the start and end values, not every element) and _sometimes_ have a fast path for methods like `#include?` if you get lucky with the types of the start and end values, but for small arrays and/or ranges where the bounds aren’t the right types for the fast path, it doesn’t make a practical difference.